master/internal/task/allocation.go

```diff
@@ -0,0 +1,405 @@
+// ensures we don't leave any resources running.
+func (a *Allocation) Cleanup(ctx *actor.Context) {
+	if err := a.db.DeleteAllocationSession(a.model.AllocationID); err != nil {
+		ctx.Log().WithError(err).Error("error delete allocation session")
+	}
+	if err := a.db.CompleteAllocation(&a.model); err != nil {
+		ctx.Log().WithError(err).Error("failed to mark allocation completed")
+	}
+	// Just in-case code.
+	if !a.exited {
+		ctx.Log().Info("exit did not run properly")
+		for cID, r := range a.reservations {
+			if _, ok := a.exitedContainers[cID]; !ok {
+				ctx.Log().Infof("allocation exited with unterminated reservation: %v", r.Summary())
+				r.Kill(ctx)
+			}
+		}
+		ctx.Tell(a.rm, sproto.ResourcesReleased{TaskActor: ctx.Self()})
+	}
+}
+
+// ResourcesAllocated handles receiving resources from the resource manager. Note: it makes a single
+// ask to the parent to build its task spec.. this is mostly a hack to defer lots of computationally
+// heavy stuff unless it is necessarily (which also works to spread occurrences of the same work
+// out). Eventually, Allocations should just be started with their TaskSpec.
+func (a *Allocation) ResourcesAllocated(ctx *actor.Context, msg sproto.ResourcesAllocated) error {
+	// Get the task spec first, so the trial/task table is populated before allocations.
+	resp := ctx.Ask(ctx.Self().Parent(), BuildTaskSpec{})
+	if err := resp.Error(); err != nil {
+		return errors.Wrapf(err, "could not get task spec")
+	}
+	spec := *resp.Get().(*tasks.TaskSpec)
+
+	for _, r := range msg.Reservations {
+		a.reservations[r.Summary().ID] = r
+	}
+
+	if err := a.db.AddAllocation(&a.model); err != nil {
+		return errors.Wrap(err, "saving trial allocation")
+	}
+
+	token, err := a.db.StartAllocationSession(a.model.AllocationID)
+	if err != nil {
+		return errors.Wrap(err, "starting a new allocation session")
+	}
+
+	a.preemption = NewPreemption(a.model.AllocationID)
+	a.rendezvous = NewRendezvous(a.model.AllocationID, ranksFromReservations(msg.Reservations))
+	for cID, r := range a.reservations {
+		r.Start(ctx, spec, sproto.ReservationRuntimeInfo{
+			Token:        token,
+			AgentRank:    a.rendezvous.rank(cID),
+			IsMultiAgent: len(a.reservations) > 1,
+		})
+	}
+	actors.NotifyAfter(ctx, RendezvousTimeoutDuration, RendezvousTimeout{
+		AllocationID: a.model.AllocationID,
+	})
+	return nil
+}
+
+// SetReservationAsDaemon sets the reservation as a daemon reservation. This means we won't wait for
+// it to exit in errorless exits and instead will kill the forcibly.
+func (a *Allocation) SetReservationAsDaemon(
+	ctx *actor.Context, aID model.AllocationID, cID cproto.ID,
+) {
+	var err error
+	if aID != a.model.AllocationID {
+		err = ErrStaleAllocation{aID, a.model.AllocationID}
+	} else if _, ok := a.reservations[cID]; !ok {
+		err = ErrStaleContainer{ID: cID}
+	}
+	if err != nil {
+		ctx.Respond(err)
+		return
+	}
+
+	a.daemonReservations[cID] = true
```

Comment:
what if all containers get marked as daemon containers?  I don't see a check for that